Add stats for `query` and `cache` requests

GKosheev opened the pull request:
Issue #64  

I implemented ```get``` method to get the size of cached calls: 

```javascript
get numberOfCachedCalls() {
    return this._cache.size
}
```

szmarczak commented:
This is actually incorrect. It doesn't return the total number of cached requests, but the current cache size. I'd implement it like this:

```js
this.stats = {
	cache: 0,
	query: 0,
};
```

and increment the counters respectively in https://github.com/szmarczak/cacheable-lookup/blob/5b08a3610a5ec1328da583a6327111428ff85c42/source/index.js#L213

You also need to provide some tests.

GKosheev commented:
In the last commit, I erased created method ```numberOfCachedCalls``` and added counters to ```query``` method.

Also, I implemented unit test: 
```javascript
test('cache and query stats', async t => {
	const cacheable = new CacheableLookup();
	t.is(cacheable.stats.query, 0);
	t.is(cacheable.stats.cache, 0);

	await cacheable.lookupAsync('1dot1dot1dot1.cloudflare-dns.com');
	t.is(cacheable.stats.query, 1);
	t.is(cacheable.stats.cache, 1);

	await cacheable.lookupAsync('1dot1dot1dot1.cloudflare-dns.com');
	t.is(cacheable.stats.query, 2);
	t.is(cacheable.stats.cache, 1);

	await cacheable.lookupAsync('google.com');
	t.is(cacheable.stats.query, 3);
	t.is(cacheable.stats.cache, 2);
});

```

szmarczak commented:
@GKosheev No need to duplicate code in the comments :) There's a `review` feature and you can add comments there if necessary.

szmarczak commented on source/index.js:
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
 	}
 
 	async query(hostname) {
+		this.stats.query++;
```
This is incorrect. `query` should be only incremented when `source` is set to `query`.

szmarczak replied:
```suggestion
```

szmarczak commented on source/index.js:
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
 				const newPromise = this.queryAndCache(hostname);
 				this._pending[hostname] = newPromise;
-
+				this.stats.cache++;
```
source is set to `query` but you increment `cache` instead.

szmarczak replied:
```suggestion
				this.stats.query++;
```

szmarczak commented on source/index.js:
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
 			const pending = this._pending[hostname];
-
 			if (pending) {
 				cached = await pending;
```
```suggestion
				this.stats.cache++;
				cached = await pending;
```

szmarczak commented on source/index.js:
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
 		let source = 'cache';
 		let cached = await this._cache.get(hostname);
 
 		if (!cached) {
```
```suggestion
		if (cached) {
			this.stats.cache++;
		} else {
```

GKosheev replied:
I don't think I understand this particular suggestion quite well on top of others. 
Do you want me to get rid of ```!``` and make a different ``` if``` statement?

szmarczak commented on tests/test.js:
```diff
@@ -1067,3 +1067,21 @@ test.cb.failing('throws original lookup error if not recognized', t => {
 });
+
+test('cache and query stats', async t => {
+	const cacheable = new CacheableLookup();
```
Can you use the mocked resolver like in the tests above?

GKosheev replied:
I'm trying to understand how to use ```resolver```. 
Would it be enough if I just add it to the existing test like that?
```javascript
const cacheable = new CacheableLookup({resolver});
```

szmarczak replied:
Here's an example: https://github.com/GKosheev/cacheable-lookup/blob/ca90ec3857e219af14c7ec7c2b9580c5de5f2601/tests/test.js#L408-L430

GKosheev commented:
I'm not sure if I did the unit test as you wanted me to, I tried to follow the example you sent me with the ```verify``` method.  
But I am always ready to make corrections.

szmarczak commented on tests/test.js:
```diff
@@ -1067,3 +1067,33 @@ test.cb.failing('throws original lookup error if not recognized', t => {
+test('cache and query stats', async t => {
+	const cacheable = new CacheableLookup({resolver});
+
+	let entries = await cacheable.lookupAsync('1dot1dot1dot1.cloudflare-dns.com', {all: true, family: 4});
```
Please use a hostname present in the mocked resolver

szmarczak commented on tests/test.js:
```diff
@@ -1067,3 +1067,33 @@ test.cb.failing('throws original lookup error if not recognized', t => {
+
+	let entries = await cacheable.lookupAsync('1dot1dot1dot1.cloudflare-dns.com', {all: true, family: 4});
+	verify(t, entries, [
+		{family: 4, source: 'query'}
```
Missing `address`

GKosheev commented:
In the last commit, I added the missing ```address``` and changed the hostname to ```temporary```.

GKosheev commented:
@szmarczak, would that work?

szmarczak commented:
Will take a look now, sorry for delay. Also, no need to post a comment of your commit changes :)

szmarczak commented:
Looks good! Thanks for contributing! :heart:

wichopy commented:
@szmarczak @GKosheev I dont see an api in the docs to access the stats, do I just do cacheable.stats ??
